### PR TITLE
Fully const layouts

### PR DESCRIPTION
**Pull request by Skelebot**
This commit introduces fully const layouts - instead of
```rust
pub type Layers<T = core::convert::Infallible> = &'static [&'static [&'static [Action<T>]]];
```
now there's
```rust
pub type Layers<T, const C: usize, const R: usize, const L: usize> = [[[Action<T>; C]; R]; L];
```
This makes sure that the layout's layer count, row count and keycode count matches exactly what the user specified; protects against errors when too many or too little keycodes were specified in the layout, resulting in weird runtime behaviour (it even went unnoticed in line 40 in src/layout.rs, which also gets fixed in this PR), but also simplifies layout declaration a little (as seen in tests).

The only issue for now is that [default parameters with const generics](https://github.com/rust-lang/rust/issues/44580) are not yet supported (on stable) - meaning that the `T` in `Action<T>` has to be always specified. The simplest solution that I found is aliasing `core::convert::Infallible` to a name like `NoCustom`, which means that when the user doesn't need any custom actions, the layout declaration looks like this:
```rust
let layers = Layers<NoCustom, 12, 6, 2> = (...)
```
This also holds back some of rust's automatic type inference, meaning that unless custom actions *are* used in the layout, type annotations are needed to infer `T` (here `NoCustom`).
Obviously when defaults with const generics are stabilized, this problem can be removed without any breaking changes.

### CODE REVIEW

**Comment by gilescope** [comment]
I’m fine with it.

**Comment by TeXitoi** [comment]
> This is an issue only on perfectly rectangular ortho keyboards (which make use of every single button that on their matrix)

I'm not sure I understand your viewpoint: the size of the layout is not related to the size of the GPIO matrix, you can have different size for each. Also, with the chord features, you'll generate virtual switches presses, I suppose the user will add a dedicated row to their layout to handle them.

Your proposition only force the layout matrix to have all their row of the same size, no?

**Comment by TeXitoi** [comment]
OK, every layer has the same qmout of key is an argument.

Note: out of bound just NoOp, no panic is possible in this case.

**Comment by mriise** [comment]
const generic defaults are now in beta :D 

https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=7cd7c8f93f875d9475f0c938bee3b9aa

**Comment by TeXitoi** [comment]
https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=1088241b4402d391947db92162a3113e ils more what we want.

File: src/matrix.rs [hunk 1]
```diff
@@ -16,6 +24,8 @@ where
     C: InputPin,
     R: OutputPin,
 {
+    // Creates a new Matrix; assumes columns are pull-up inputs,
+    // and rows are output pins which are set high when not being scanned.
```
Comment (TeXitoi) [inline src/matrix.rs:28]:
```suggestion
    /// Creates a new Matrix.
    /// 
    /// Assumes columns are pull-up inputs,
    /// and rows are output pins which are set high when not being scanned.
```

File: src/matrix.rs [hunk 2]
```diff
@@ -35,6 +45,9 @@ where
         }
         Ok(())
     }
+    // Scans the matrix and checks which keys are pressed.
+    // Every row pin in order is pulled low, and then each column
+    // pin is tested; if it's low, the key is marked as pressed.
```
Comment (TeXitoi) [inline src/matrix.rs:50]:
```suggestion
    /// Scans the matrix and checks which keys are pressed.
    ///
    /// Every row pin in order is pulled low, and then each column
    /// pin is tested; if it's low, the key is marked as pressed.
```

File: src/matrix.rs [hunk 3]
```diff
@@ -2,6 +2,14 @@
 
 use embedded_hal::digital::v2::{InputPin, OutputPin};
 
+/// Describes the hardware-level matrix of switches.
```
Comment (TeXitoi) [inline src/matrix.rs:5]:
```suggestion
/// Describes the hardware-level matrix of switches.
///
```

**Review by TeXitoi — APPROVED** [review]
Thanks, and sorry for the late response.

**Pull request merged**